src/mark.c

```diff
@@ -203,55 +206,87 @@ gboolean mark_run(const gchar *state,
 	g_assert(message != NULL && *message == NULL);
 
-	slot = get_slot_by_identifier(slot_identifier, &ierror);
+	slots = get_slots_by_identifier(slot_identifier, &ierror);
 	if (ierror) {
-		res = FALSE;
 		*message = g_strdup(ierror->message);
-		goto out;
+		return FALSE;
 	}
 
 	if (g_strcmp0(state, "good") == 0) {
+		if (g_list_length(slots) > 1) {
+			*message = g_strdup("cannot mark multiple slots as good");
+			if (slot_name)
+				*slot_name = g_strdup("other");
+			return FALSE;
+		}
+		RaucSlot *slot = (RaucSlot*)g_list_first(slots)->data;
+
 		if (!r_mark_good(slot, &ierror)) {
-			res = FALSE;
 			*message = g_strdup(ierror->message);
-		} else {
-			res = TRUE;
-			*message = g_strdup_printf("marked slot %s as good", slot->name);
+			return FALSE;
+		}
+
+		if (slot_name)
+			*slot_name = g_strdup(slot->name);
+		*message = g_strdup_printf("marked slot %s as good", slot->name);
+
+		if (r_context()->config->prevent_fallback
+		    && g_strcmp0(slot_identifier, "booted") == 0) {
+			g_autofree gchar *mark_bad_message = NULL;
+
+			if(!mark_run("bad", "other", NULL, &mark_bad_message)) {
+				*message = g_strdup_printf("%s, but failed with: %s", *message, mark_bad_message);
+				return FALSE;
+			}
+
+			*message = g_strdup_printf("%s and also %s", *message, mark_bad_message);
 		}
 	} else if (g_strcmp0(state, "bad") == 0) {
-		if (!r_mark_bad(slot, &ierror)) {
-			res = FALSE;
-			*message = g_strdup(ierror->message);
-		} else {
-			res = TRUE;
-			*message = g_strdup_printf("marked slot %s as bad", slot->name);
+		g_autoptr(GString) slot_names = g_string_new(NULL);
+
+		for (GList *l = slots; l != NULL; l = l->next) {
+			RaucSlot *slot = l->data;
+			if (!r_mark_bad(slot, &ierror)) {
+				*message = g_strdup(ierror->message);
+				if (slot_name)
+					*slot_name = g_strdup(slot->name);
+				return FALSE;
+			}
+			g_string_append_printf(slot_names, "%s ", slot->name);
 		}
+
+		if (slot_names->len > 1)
+			g_string_truncate(slot_names, slot_names->len - 1);
+
+		*message = g_strdup_printf("marked slot[s] %s as bad", slot_names->str);
+		if (slot_name)
+			*slot_name = g_strdup("other");
 	} else if (g_strcmp0(state, "active") == 0) {
+		if (g_list_length(slots) > 1) {
+			*message = g_strdup("cannot activate multiple slots");
+			if (slot_name)
+				*slot_name = g_strdup("other");
+			return FALSE;
+		}
+		RaucSlot *slot = (RaucSlot*)g_list_first(slots)->data;
+
+		if (slot_name)
+			*slot_name = g_strdup(slot->name);
+
 		if (!r_mark_active(slot, &ierror)) {
-			res = FALSE;
 			*message = g_strdup(ierror->message);
-		} else {
-			res = TRUE;
-			*message = g_strdup_printf("activated slot %s", slot->name);
+			return FALSE;
 		}
-		g_clear_error(&ierror);
+		*message = g_strdup_printf("activated slot %s", slot->name);
 	} else {
-		res = FALSE;
 		*message = g_strdup_printf("unknown subcommand %s", state);
+		return FALSE;
 	}
-
-out:
-	if (res && slot_name)
-		*slot_name = g_strdup(slot->name);
```

Comment:
With the new code, we only get here in the successful case. As `slot_name` should be set only then, you should be able to reduce complexity and code duplication by generating the space-separated string here only once at the end. That would move that part of the loop code in the "bad" case here.